### PR TITLE
Spanish Chapter 3: sections 1 & 2

### PR DESCRIPTION
**Pull request by fordaz**
Related to https://github.com/huggingface/course/issues/38

### CODE REVIEW

**Comment by lgesuellip** [comment]
Hi,

I have been working in these sections so, If it is ok with you, I could review that and check if we have any differences.

Thanks,

**Comment by fordaz** [comment]
@lewtun thanks for pointing this out, I'll update the _toctree.yml as well.
Regarding the translation of Fine-tuning, I considered both "Ajustar" y "Entonar", but I can see how preserving the original term in English might be better in this case. I can use that and see how other reviewers see it.

**Comment by fordaz** [comment]
@lgesuellip I'd be glad to collaborate with you. Please check and compare this with your translation, and feel free to suggest updates/improvements in the PR. Thanks !

**Comment by lewtun** [comment]
> Regarding the translation of Fine-tuning, I considered both "Ajustar" y "Entonar", but I can see how preserving the original term in English might be better in this case. I can use that and see how other reviewers see it.

Using the English term for these technical words sounds good to me!

**Comment by fordaz** [comment]
@lewtun I noticed few things:

* Translations of chapter 1 (sections 1,3,4,8) are using "Ajustar". It might be nice to have a consistent translations of these terms.

* The Spanglish of statements like this "Como usar la API de alto nivel del entrenador para fine-tune un modelo." did not feel right, just a personal impression.

At this point, I'm open to ideas, but personally I'm leaning towards using "Ajustar". I'm also using an extra hint in the toc "Ajuste (fine-tuning) de un modelo preentrenado" so folks have a reference of what "ajustar" means.

**Comment by lewtun** [comment]
> At this point, I'm open to ideas, but personally I'm leaning towards using "Ajustar".

Hehe, yes translation is _hard_ 😅. Since we already have "ajustar" in the previous chapters, let's keep that format for now. I like your idea to include the English phrase, so one idea would be to create a `glossary` chapter that provides a vocabulary table - this is what has been done in e.g. the Persian and German translations: https://github.com/huggingface/course/blob/main/chapters/de/glossary/1.mdx

Similarly one could have a `TRANSLATING.txt` file that provides a guide to other translators. Again the Persian team have created a template that was adapted to the German case here: https://github.com/huggingface/course/blob/main/chapters/de/TRANSLATING.txt

No need to handle these issues in this PR - just an idea for the future if you have time / interest :)

**Comment by fordaz** [comment]
@lewtun I was actually thinking about a glossary, I really like the idea of both files, this will help the translators to be consistent. I'll work on it on a separate PR. Thanks for your input.

**Comment by fordaz** [comment]
Hi @lewtun, I have a first version of `es/TRANSLATING.txt` and `es/glossary/1.mdx` (an adaptation of the German files). Do you suggest to add them as part of this PR or a separate one ? Thanks.

**Review by lgesuellip — CHANGES_REQUESTED** [review]

File: chapters/es/chapter3/2.mdx [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+
+### Preprocesando un conjunto de datos
+
+{#if fw === 'pt'}
+<Youtube id="0u3ioSwev3s"/>
+{:else}
+<Youtube id="P-rZWqcB6CE"/>
+{/if}
+
+Para preprocesar el conjunto de datos, necesitamos convertir el texto en números que puedan ser entendidos por el modelo. Como viste en el [capítulo anterior](/course/chapter2), esto se hace con el toquenizador. Podemos darle al tokenizador una oración o una lista de oraciones, así podemos tokenizar directamente todas las primeras oraciones y todas segundas oraciones de cada par de la siquiente manera:
```
Comment (lgesuellip) [inline chapters/es/chapter3/2.mdx:162]: Para preprocesar el conjunto de datos, necesitamos convertir el texto en números que puedan ser entendidos por el modelo. Como viste en el [capítulo anterior](/course/chapter2), esto se hace con el **tokenizador**. Podemos darle al tokenizador una oración o una lista de oraciones, así podemos tokenizar directamente todas **las primeras  y las segundas** oraciones de cada par de la **siguiente** manera.

File: chapters/es/chapter3/2.mdx [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+from transformers import AutoTokenizer
+
+checkpoint = "bert-base-uncased"
+tokenizer = AutoTokenizer.from_pretrained(checkpoint)
+tokenized_sentences_1 = tokenizer(raw_datasets["train"]["sentence1"])
+tokenized_sentences_2 = tokenizer(raw_datasets["train"]["sentence2"])
+```
+
+Sin embargo, no podemos simplemente pasar dos secuencias al modelo y obtener una predicción indicando si estas son parafrases o no. Necesitamos manipular las dos secuencias como un par y aplicar el preprocesamiento apropiado.
+Afortunadamente, el tonenizador también puede tomar un par de secuencias y preparalo de la manera como nuestro modelo BERT lo espera:
```
Comment (lgesuellip) [inline chapters/es/chapter3/2.mdx:174]: I wrote: Afortunadamente, el **tokenizador** puede recibir también un par de oraciones y **preparar** las misma de una forma que nuestro modelo BERT espera.

**Review by lgesuellip — CHANGES_REQUESTED** [review]

File: chapters/es/chapter3/1.mdx [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+# Introducción
+
+En el [Capítulo 2](/course/chapter2) exploramos como usar los tokenizadores y modelos preentrenados para hacer predicciones. Pero qué tal si deseas ajustar un modelo preentrenado con tu propio conjunto de datos ?
+
+{#if fw === 'pt'}
+* Como preparar un conjunto de datos grande desde el Hub.
+* Como usar la API de alto nivel del entrenador para ajustar un modelo.
+* Como usar un lazo personalizado de entrenamiento.
+* Como aprovechar la libreria acelerada 🤗 para fácilmente ejecutar el lazo personalizado de entrenamiento en cualquier configuración distribuida.
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/1.mdx:11]: good point. I'll update it as `la Accelerate library 🤗 `.

File: chapters/es/chapter3/2.mdx [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+  classNames="absolute z-10 right-0 top-0"
+  options={[
+    {label: "Google Colab", value: "https://colab.research.google.com/github/huggingface/notebooks/blob/master/course/chapter3/section2_tf.ipynb"},
+    {label: "Aws Studio", value: "https://studiolab.sagemaker.aws/import/github/huggingface/notebooks/blob/master/course/chapter3/section2_tf.ipynb"},
+]} />
+
+{/if}
+
+{#if fw === 'pt'}
+Continuando con el ejemplo del [capítulo anterior](/course/chapter2), aquí mostramos como podríamos entrenar un clasificador secuencial en un lote en PyTorch:
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:26]: Good catch !!

File: chapters/es/chapter3/2.mdx [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+# This is new
+model.compile(optimizer="adam", loss="sparse_categorical_crossentropy")
+labels = tf.convert_to_tensor([1, 1])
+model.train_on_batch(batch, labels)
+```
+{/if}
+
+Por supuesto, entrenando el modelo con solo dos oraciones no va a producir muy buenos resultados. Para obtener mejores resultados, debes preparar un conjunto de datos mas grande.
+
+En esta sección usaremos como ejemplo el conjunto de datos MRPC (Cuerpo de parafrases de investigaciones de Microsoft), que fue presentado en el [artículo](https://www.aclweb.org/anthology/I05-5002.pdf) de William B. Dolan and Chris Brockett. El conjunto de datos consiste en 5,801 pares of oraciones, con una etiqueta que indica si son parafrases o no. (es decir, si ambas oraciones significan lo mismo). Este lo hemos seleccionado para este capítulo porque es un conjunto de datos pequeño que facilita la experimentación y entrenamiento sobre él.
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:77]: Good catch !

[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:77]: Like it!

File: chapters/es/chapter3/2.mdx [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+### Cargando un conjunto de datos desde el Hub
+
+{#if fw === 'pt'}
+<Youtube id="_BZearw7f0w"/>
+{:else}
+<Youtube id="W_gMJF0xomE"/>
+{/if}
+
+El Hub no solo contiene modelos; sino que también tiene múltiples conjunto de datos en muchos lenguajes diferentes. Puedes explorar los conjuntos de datos [aquí](https://huggingface.co/datasets), y recomendamos que trates de cargar y procesar un nuevo conjunto de datos una vez que hayas revisado esta sección (mira la documentación general [aquí](https://huggingface.co/docs/datasets/loading_datasets.html#from-the-huggingface-hub)). Por ahora, enfoquémonos en el conjunto de datos MRPC! Este es uno de los 10 conjuntos de datos que comprende el [punto de referencia GLUE](https://gluebenchmark.com/), el cual es un punto de referencia académico que se usa para medir el desempeño de modelos ML sobre 10 tareas de clasificación de texto.
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:87]: got it.

File: chapters/es/chapter3/2.mdx [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+    test: Dataset({
+        features: ['sentence1', 'sentence2', 'label', 'idx'],
+        num_rows: 1725
+    })
+})
+```
+
+Como puedes ver, obtenemos un objeto `DatasetDict` que contiene los conjuntos de datos de entrenamiento, de validación y de pruebas. Cada uno de estos contiene varias columnas (`sentence1`, `sentence2`, `label`, and `idx`) y un número variable de filas, que son el número de elementos en cada conjunto (asi, que hay 3,668 pares de oraciones en el conjunto de entrenamiento, 408 en el de validación, y 1,725 en el pruebas)
+
+Este comando descarga y memoriza el conjunto de datos, por defecto en *~/.cache/huggingface/dataset*. Recuerda del Capítulo 2 que puedes personalizar tu carpeta mediante la configuración de la variable de ambiente `HF_HOME`.
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:117]: Good point. I was not happy with it either. I even thought about keeping cache. But `almacena` sounds good.

[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:117]: Like it.

File: chapters/es/chapter3/2.mdx [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+
+Podemos acceder cada par de oraciones en nuestro objeto `raw_datasets` usando indexación, como con un diccionario.
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:119]: got it.

File: chapters/es/chapter3/2.mdx [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+```
+
+```python out
+{'idx': 0,
+ 'label': 1,
+ 'sentence1': 'Amrozi accused his brother , whom he called " the witness " , of deliberately distorting his evidence .',
+ 'sentence2': 'Referring to him as only " the witness " , Amrozi accused his brother of deliberately distorting his evidence .'}
+```
+
+Podemos ver que las etiquetas ya son números enteros, así que no es necesario hacer ningún preprocesamiento. Para saber cual valor corresponde con cual etiqueta, podemos inspeccionar el campo `features` de nuestro `raw_train_dataset`. Esto indicara el tipo de cada column:
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:133]: Got it.

[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:133]: Got it.

File: chapters/es/chapter3/2.mdx [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+```
+
+```python out
+{'sentence1': Value(dtype='string', id=None),
+ 'sentence2': Value(dtype='string', id=None),
+ 'label': ClassLabel(num_classes=2, names=['not_equivalent', 'equivalent'], names_file=None, id=None),
+ 'idx': Value(dtype='int32', id=None)}
+```
+
+Internamente, `label` es de tipo `ClassLabel`, y la asociación de valores enteros y sus etiquetas esta almacenado en la carpeta *names*. `0` corresponde con `not_equivalent`, y `1` corresponde con `equivalent`.
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:146]: Got it.

File: chapters/es/chapter3/2.mdx (hunk 1 above)
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:162]: Good catch !

File: chapters/es/chapter3/2.mdx (hunk 2 above)
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:174]: Like it.

**Comment by fordaz** [comment]
Great feedback @lgesuellip ! Thanks a lot !

File: chapters/es/chapter3/2.mdx [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+
+{#if fw === 'pt'}
+<Youtube id="_BZearw7f0w"/>
+{:else}
+<Youtube id="W_gMJF0xomE"/>
+{/if}
+
+El Hub no solo contiene modelos; sino que también tiene múltiples conjunto de datos en muchos lenguajes diferentes. Puedes explorar los conjuntos de datos [aquí](https://huggingface.co/datasets), y recomendamos que trates de cargar y procesar un nuevo conjunto de datos una vez que hayas revisado esta sección (mira la documentación general [aquí](https://huggingface.co/docs/datasets/loading_datasets.html#from-the-huggingface-hub)). Por ahora, enfoquémonos en el conjunto de datos MRPC! Este es uno de los 10 conjuntos de datos que comprende el [punto de referencia GLUE](https://gluebenchmark.com/), el cual es un punto de referencia académico que se usa para medir el desempeño de modelos ML sobre 10 tareas de clasificación de texto.
+
+La Libreria de Conjunto de datos 🤗 provee un comando muy simple para descargar y memorizar un conjunto de datos en el Hub. Podemos descargar el conjunto de datos de esta manera:
```
[the remark this answers is not in the document]
Comment (lgesuellip) [reply chapters/es/chapter3/2.mdx]: I like better the emoji after the name, it makes sense.

**Comment by lgesuellip** [comment]
Excellent work @forda, the translation looks fluently and nice to read 📖✨

**Review by lgesuellip — CHANGES_REQUESTED** [review]

File: chapters/es/chapter3/2.mdx [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+# This is new
+batch["labels"] = torch.tensor([1, 1])
+
+optimizer = AdamW(model.parameters())
+loss = model(**batch).loss
+loss.backward()
+optimizer.step()
+```
+{:else}
+Continuando con el ejemplo del [capítulo anterior](/course/chapter2), aquí mostramos como podríamos entrenar un clasificador secuencial en un lote en TensorFlow:
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:51]: got it.

File: chapters/es/chapter3/2.mdx [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+```
+
+```python out
+{'idx': 0,
+ 'label': 1,
+ 'sentence1': 'Amrozi accused his brother , whom he called " the witness " , of deliberately distorting his evidence .',
+ 'sentence2': 'Referring to him as only " the witness " , Amrozi accused his brother of deliberately distorting his evidence .'}
+```
+
+Podemos ver que las etiquetas ya son números enteros, así que no es necesario hacer ningún preprocesamiento. Para saber cual valor corresponde con cual etiqueta, podemos inspeccionar el atributo `features` de nuestro `raw_train_dataset`. Esto indicara el tipo dato de cada column:
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:133]: done.

File: chapters/es/chapter3/2.mdx [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+
+```python out
+{ 
+  'input_ids': [101, 2023, 2003, 1996, 2034, 6251, 1012, 102, 2023, 2003, 1996, 2117, 2028, 1012, 102],
+  'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1],
+  'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
+}
+```
+
+Nosotros consideramos las llaves `input_ids` y `attention_mask` en el [Capítulo 2](/course/chapter2), per nos saltamos la llave `token_type_ids`. En este ejemplo, esta es la que le dice al modelo cual parte de la entrada es la primera oración y cual es la segunda.
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:189]: oh yeah, like it.

File: chapters/es/chapter3/2.mdx [hunk 15]
```diff
@@ -0,0 +1,384 @@
...
+```python out
+['[CLS]', 'this', 'is', 'the', 'first', 'sentence', '.', '[SEP]', 'this', 'is', 'the', 'second', 'one', '.', '[SEP]']
+[      0,      0,    0,     0,       0,          0,   0,       0,      1,    1,     1,        1,     1,   1,       1]
+```
+
+Como puedes observar, las partes de la entrada que corresponden a `[CLS] sentence1 [SEP]` todas tienen un tipo de token ID `0`, mientras que las otras partes que corresponden a `sentence2 [SEP]`, todas tienen tipo ID `1`.
+
+Nótese que si seleccionas un punto de control diferente, no necesariamente tendrás el `token_type_ids` en tus entradas tonenizadas (por ejemplo, ellas no aparecen si usas un modelo DistilBERT). Estas aparecen cuando el modelo sabe que hacer con ellas, porque las ha visto durante su etapa de preentrenamiento.
+
+Aquí, BERT esta preentrenado con tokens de tipo ID, y además del objetivo de modelado de lenguaje oculto que mencionamos en el [Capítulo 1](/course/chapter1), también tiene el objectivo llamado _predicción de la siguiente oración_. El objectivo con esta tarea es modelar la relación entre pares de oraciones.
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:221]: got it.

File: chapters/es/chapter3/2.mdx [hunk 16]
```diff
@@ -0,0 +1,384 @@
...
+```py
+tokenized_dataset = tokenizer(
+    raw_datasets["train"]["sentence1"],
+    raw_datasets["train"]["sentence2"],
+    padding=True,
+    truncation=True,
+)
+```
+
+Esto funciona bien, pero tiene la desventaja de que devuelve un diccionario (con nuestras llaves, `input_ids`, `attention_mask`, and `token_type_ids`, y valores que son listas de listas). Además va a trabajar solo si tienes suficiente memoria principal para almacenar todo el conjunto de datos durante la tokenización (mientras que los conjuntos de datos de la libreria 🤗 son archivos [Apache Arrow](https://arrow.apache.org/) almacenados en disco, y así solo mantienes en memoria las muestras que necesitas).
```
[the remark this answers is not in the document]
Comment (fordaz) [reply chapters/es/chapter3/2.mdx:238]: good catch.

**Review by lgesuellip — APPROVED** [review]

**Comment by fordaz** [comment]
Hi @lewtun, please let me know if any further changes are needed. I will be glad to work on them. Thanks.

**Review by lewtun — APPROVED** [review]
Thanks for iterating @fordaz - this looks fantastic 🔥 !

**Pull request merged**